File: iosApp/iosApp/Modules/MnemonicNew/MnemonicNewWireframeFactory.swift
// Created by web3d3v on 11/02/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import UIKit
import web3walletcore

// MARK: - MnemonicNewWireframeFactory

protocol MnemonicNewWireframeFactory {
    func make(
        _ parent: UIViewController?,
        context: MnemonicNewWireframeContext
    ) -> MnemonicNewWireframe
}

// MARK: - DefaultMnemonicNewWireframeFactory

final class DefaultMnemonicNewWireframeFactory {
    private let signerStoreService: SignerStoreService
    private let passwordService: PasswordService
    private let clipboardService: ClipboardService
    private let settingsService: SettingsService
    private let addressService: AddressService
    
    init(
        signerStoreService: SignerStoreService,
        passwordService: PasswordService,
        clipboardService: ClipboardService,
        settingsService: SettingsService,
        addressService: AddressService
    ) {
        self.signerStoreService = signerStoreService
        self.passwordService = passwordService
        self.clipboardService = clipboardService
        self.settingsService = settingsService
        self.addressService = addressService
    }
}

extension DefaultMnemonicNewWireframeFactory: MnemonicNewWireframeFactory {

    func make(
        _ parent: UIViewController?,
        context: MnemonicNewWireframeContext
    ) -> MnemonicNewWireframe {
        DefaultMnemonicNewWireframe(
            parent,
            context: context,
            signerStoreService: signerStoreService,
            passwordService: passwordService,
            clipboardService: clipboardService,
            settingsService: settingsService,
            addressService: addressService
        )
    }
}

// MARK: - Assembler

final class MnemonicNewWireframeFactoryAssembler: AssemblerComponent {
    
    func register(to registry: AssemblerRegistry) {
        registry.register(scope: .instance) { resolver -> MnemonicNewWireframeFactory in
            DefaultMnemonicNewWireframeFactory(
                signerStoreService: resolver.resolve(),
                passwordService: resolver.resolve(),
                clipboardService: resolver.resolve(),
                settingsService: resolver.resolve(),
                addressService: resolver.resolve()
            )
        }
    }
}

File: androidApp/src/main/java/com/sonsofcrypto/web3wallet/android/modules/compose/mnemonicnew/MnemonicNewWireframeFactory.kt
package com.sonsofcrypto.web3wallet.android.modules.compose.mnemonicnew

import androidx.fragment.app.Fragment
import com.sonsofcrypto.web3lib.services.keyStore.KeyStoreService
import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3wallet.android.common.AssemblerComponent
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistry
import com.sonsofcrypto.web3wallet.android.common.AssemblerRegistryScope
import com.sonsofcrypto.web3walletcore.modules.mnemonicNew.MnemonicNewWireframe
import com.sonsofcrypto.web3walletcore.modules.mnemonicNew.MnemonicNewWireframeContext
import com.sonsofcrypto.web3walletcore.services.password.PasswordService
import smartadapter.internal.extension.name

interface MnemonicNewWireframeFactory {
    fun make(parent: Fragment?, context: MnemonicNewWireframeContext): MnemonicNewWireframe
}

class DefaultMnemonicNewWireframeFactory(
    private val keyStoreService: KeyStoreService,
    private val passwordService: PasswordService,
): MnemonicNewWireframeFactory {

    override fun make(
        parent: Fragment?, context: MnemonicNewWireframeContext
    ): MnemonicNewWireframe = DefaultMnemonicNewWireframe(
        parent?.let { WeakRef(it) },
        context,
        keyStoreService,
        passwordService,
    )
}

class MnemonicNewWireframeFactoryAssembler: AssemblerComponent {

    override fun register(to: AssemblerRegistry) {

        to.register(MnemonicNewWireframeFactory::class.name, AssemblerRegistryScope.INSTANCE) {
            DefaultMnemonicNewWireframeFactory(
                it.resolve(KeyStoreService::class.name),
                it.resolve(PasswordService::class.name),
            )
        }
    }
}